api/s2n.h

```diff
@@ -77,6 +77,22 @@ extern int s2n_connection_set_fd(struct s2n_connection *conn, int readfd);
 extern int s2n_connection_set_read_fd(struct s2n_connection *conn, int readfd);
 extern int s2n_connection_set_write_fd(struct s2n_connection *conn, int readfd);
 
+/* Callback to send or receive data that may be defined by the user. These
+ * may be blocking or nonblocking. 
+ *
+ * These callbacks take as input a context containing anything needed in the 
+ * function (for example, a file descriptor), the buffer holding data to be sent 
+ * or received, and the length of the buffer. 
+ *
+ * The callback may send or receive less than the requested length. The function 
+ * should return the number of bytes sent/received or an error code < 0. */
+typedef int s2n_connection_send(void *io_context, const uint8_t *buf, uint32_t len);
+typedef int s2n_connection_recv(void *io_context, uint8_t *buf, uint32_t len);
+extern int s2n_connection_set_recv_ctx(struct s2n_connection *conn, void *ctx);
+extern int s2n_connection_set_send_ctx(struct s2n_connection *conn, void *ctx);
+extern int s2n_connection_set_recv_cb(struct s2n_connection *conn, s2n_connection_recv recv);
+extern int s2n_connection_set_send_cb(struct s2n_connection *conn, s2n_connection_send send);
+
```

Comment:
I flipped back-and-forth about naming these send/recv or read/write. They more closely resemble the read/write method signatures, but I stuck with send/recv to match convention (for example in `s2n_stuffer_recv_from_fd`). Open to suggestions here.